### PR TITLE
Implement Objects comparison and stop trying compare values of differ…

### LINKED ISSUE #400: Api call that uses mongomock causes 500 INTERNAL SERVER ERROR but works with Mongodb
The following code causes an error when using mongomock, but not when using mongodb:

        result = self.logs_engine.LogCollection.objects(
            __raw__={'app_id': app_id, "results": {
        "$gt": [{"$size": "results"}, 0]}}).only(
            "user_input", "app_id", "results", "date_time")
        **print(result)**

The error is: TypeError: '>' not supported between instances of 'dict' and 'dict'

### PR DESCRIPTION
**Pull request by pcorpet**
…ent types. Fixes #400.

### CODE REVIEW

**Comment by pcorpet** [comment]
Also fixes #368.

File: mongomock/filtering.py [hunk 1]
```diff
@@ -167,6 +168,69 @@ def _not_nothing_and(f):
...
+
+    return op(a, b)
+
+
+def _get_compare_type(val):
+    """Get a number representing the base type of the value used for comparison.
+
+    See https://docs.mongodb.com/manual/reference/bson-type-comparison-order/
+    """
+    if isinstance(val, type(None)):
```
Comment (vmalloc) [inline mongomock/filtering.py]: Cache this through `from types import NoneType`

Comment (vmalloc) [reply mongomock/filtering.py]: Also, doesn't this enumeration exist anywhere??? Perhaps in pymongo?

File: mongomock/helpers.py [hunk 2]
```diff
@@ -269,6 +270,11 @@ def patch_datetime_awareness_in_document(value):
     # mixing tz aware and naive.
     # On top of that, MongoDB date precision is up to millisecond, where Python
     # datetime use microsecond, so we must lower the precision to mimic mongo.
+    if isinstance(value, OrderedDict):
```
Comment (vmalloc) [inline mongomock/helpers.py]: The two `if`s can be united by just initializing by the object's type: `type(obj)(...)`

**Review by vmalloc — CHANGES_REQUESTED** [review]

File: mongomock/filtering.py (hunk 1 above)
Comment (pcorpet) [reply mongomock/filtering.py]: Yep, thanks. I didn't find any Python code doing the enumeration. I've linked the mongo C++ implementation and switched to use the same ID (they have the same ordering).

File: mongomock/helpers.py (hunk 2 above)
Comment (pcorpet) [reply mongomock/helpers.py]: Smart. Thanks!

**Review by vmalloc — APPROVED** [review]

**Comment by vmalloc** [comment]
LGTM

**Pull request merged**